iron-focusables-helper.html

```diff
@@ -0,0 +1,135 @@
+<!doctype html>
+<!--
+@license
+Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+Code distributed by Google as part of the polymer project is also
+subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+-->
+<link rel="import" href="../polymer/polymer.html">
+
+<script>
+  (function() {
+    'use strict';
+
+    Polymer.IronFocusablesHelper = {
+
+      /**
+       * Returns a sorted array of tabbable nodes, including the root node.
+       * It searches the tabbable nodes in the light and shadow dom of the chidren,
+       * sorting the result by tabindex.
+       * @param {!Node} node
+       * @return {Array<Node>}
+       */
+      getTabbableNodes: function(node) {
+        var result = [];
+        // If there is at least one element with tabindex > 0, we need to sort
+        // the final array by tabindex.
+        var maxTabIndex = this._findTabbableNodes(node, result);
+        maxTabIndex > 0 && result.sort(this._tabIndexSort);
+        return result;
+      },
+
+      /**
+       * Returns if a node is focusable.
+       * @param {!Node} node
+       * @return {boolean}
+       */
+      isFocusable: function(node) {
+        // shadow roots won't have this method.
+        if (!node.hasAttribute) {
+          return false;
+        }
+        var name = node.localName;
+        // Elements that can be focused even if they have [disabled] attribute.
+        if (name === 'iframe' ||
+          ((name === 'a' || name === 'area') && node.hasAttribute('href')) ||
+          node.hasAttribute('tabindex') || node.hasAttribute('contentEditable')) {
+          return true;
+        }
+        // Elements that cannot be focused if they have [disabled] attribute.
+        return /^(input|select|textarea|button|object)$/.test(name) && !node.disabled;
+      },
+
+      /**
+       * Returns if a node is tabbable.
+       * @param {!Node} node
+       * @return {boolean}
+       */
+      isTabbable: function(node) {
+        return node.tabIndex >= 0 && this.isFocusable(node) && this._isVisible(node);
+      },
+
+      /**
+       * Searches for nodes that are tabbable and adds them to the `result` array.
+       * Returns the maximum tabindex of the added nodes.
+       * @param {!Node} node The starting point for the search; it's added if tabbable.
+       * @param {!Array<Node>} result
+       * @return {Number}
+       * @private
+       */
+      _findTabbableNodes: function(node, result) {
```

Comment:
I'm pretty sure all of those are O(n) since "O(n) + O(n)" = O(n) but, expanding out the two pieces of work (gather and max), option 1 would be "O(n) + O(n)" (maybe more clearly written as something like "n * 2O(1)" in this case) for all of those situations, even though it uses a single loop.

```javascript
function sumAndConcat1(list) {
  var sum = 0;
  var string = "";
  for (var i = 0; i < list.length; i++) {
    sum += list[i];
    string += list[i];
  }
  return {sum, string};
}


function sumAndConcat2(list) {
  var sum = 0;
  var string = "";
  for (var i = 0; i < list.length; i++) {
    sum += list[i];
  }
  for (var i = 0; i < list.length; i++) {
    string += list[i];
  }
  return {sum, string};
}
```

I don't think the use of a single for loop in `sumAndConcat1` shouldn't give it any performance advantage over `sumAndConcat2`. And, if the second loop in `sumAndConcat2` could perform fewer than `list.length` iterations (assuming the function was doing something else), wouldn't that be an advantage for `sumAndConcat2`?

However, I suppose the end to my argument wouldn't have much benefit anyway given the operation being optimized out (max) has such a small cost. :P I still think the function's signature is weird but, yeah, the rest of my argument is more pedantic than practical.